add json files generator for episodes, host, panelists and sponsors

MacKentoch opened the pull request:
## Content
#### Add JSON files generator for:
- episodes
- host
- panelists
- sponsors

Files are created/updated in **/data/json** directory.

> NOTE: 
> - **/data/json** directory must exist (_generator does not test it since it should_)
> - files could not exist or be empty 
>   - it won't matter generator will do its job
> - generator sources
>   - **generate/jsonFiles.js**
>   - **generate/writeToJsonFile.js**
>     ## task 
#### JSON generator only

`npm run build:json`
#### all validate task

`npm run validate`
## Bonus

I needed to import host data (previsouly created on the fly in **/generate/contributors**).
So host data is now in **/ressources/host/index.js** (_I updated /generate/contributors by the way_)
## To finish

Something else is coming soon...

MacKentoch committed:
add json files generator for episodes, host, panelists and sponsors

update json files generator for episodes, host, panelists and sponsors

fix lint issue

add myself as contributor

fix PR

fix files and dirs in kebab-case

fix PR updates

add createJson in build script

kentcdodds commented:
Hey @MacKentoch,
Thanks for the pull request! Looking forward to adding this. Here are a few things:
1. Let's not commit these generated files to the repo. We'll just deploy them to the website with everything else. Please add the generated files to the `.gitignore`.
2. Let's have the files go in the same directory as the `index.html` for the episode/page/etc. So we could do a GET at `https://javascriptair.com/episodes/2016-04-06/data.json` to get the episode data for that episode.
3. You made your branch right before I did a major refactor with the npm scripts. This project now uses [`p-s`](http://npm.im/p-s). Please rebase on master.

If you could make those updates, then I'll give more feedback on the implementation. Thanks!

MacKentoch commented:
@kentcdodds sure.
I will update it and follow your these points.

MacKentoch commented:
I synced my fork but I encounter build errors now related to: 
- `spread operator`  
- `destructuring syntax`.

_Using p-s or directly babel-node won't matter. It is not related._

**I don't find any issue relative to my problem so maybe I missed something?**

As a an attempt to fix:  
- I updated even my global babel, eslint... .
- I deleted `node_modules` directory and re-install

_but no luck..._

NOTE: my node and npm versions
- node: v4.2.4 
- npm: v3.8.8

---

**Spread error example:**

``` bash
/Users/kentor/GitProjects/jsairWebsite/site/generate/contributors.js:49
  return [...array, ...groupArray];
          ^^^


SyntaxError: Unexpected token ...
    at exports.runInThisContext (vm.js:53:16)
```

**Destructuring syntax error example:**

``` bash
/Users/kentor/GitProjects/jsairWebsite/site/generate/renderComponentToFile.js:21
  const { html, css } = _aphrodite.StyleSheetServer.renderStatic(() => {
        ^

SyntaxError: Unexpected token {
``
```

kentcdodds commented:
Oh, yeah, the project now requires node 6. I recommend nvm :-)

MacKentoch commented:
You caught me... Time for me to leave my prehistoric cave and use `nvm` 😄

MacKentoch commented:
**PR Updated** and it now meets your requirements.

For episodes, I create a `data.json` per episode (_in each directories `/episode/YYYY-MM-DD/`_).

But as I need a single json to fetch all episodes in a row (_I will explain you later why_), I create some kind `allEpisodes json file` (_data.json in episode root directory_).
As it could be expensive to get all episodes content in a single file; I filter episode attributes to a minimum that I needed.

Code is pretty straightforward but it is a minimum configurable (configs: `/generate/jsonFilesConfigs/`).

kentcdodds commented on generate/writeToJsonFile.js:
```diff
@@ -0,0 +1,82 @@
+function writeToFile(file, content) {
+  fs.writeFile(file, content, err => {
+    if (err) {
+      process.exit(1)
```
Should probably log the error here.

MacKentoch replied:
Since I will simplify my code to rewrite all data.json (rather updating), I won't need to read file anymore so I won't have to catch errors.

kentcdodds commented on generate/writeToJsonFile.js:
```diff
@@ -0,0 +1,82 @@
+}
+
+
+export function writeJsonArray(arrayNewItems, config, applyFilter = false) {
```
If I understand this code correctly, you're reading the existing file and only updating the changed bits. I would rather just have this override the original file with all the new data.

MacKentoch replied:
Yes that was the original aim.
I will rewrite all json files as you suggest so my code will simplify.

kentcdodds commented on generate/jsonFiles.js:
```diff
@@ -0,0 +1,26 @@
+episodes.map(
+  episode => {
+    const episodeConfig = Object.assign({}, config.episodesToJson)
+    episodeConfig.destDirectoryFromHere = `${episodeConfig.destDirectoryFromHere}${episode.date}/`
```
Should probably use `path.join` just to be clear/safe.

MacKentoch replied:
Yes you are right (what did I write... 😅)

kentcdodds commented on generate/writeToJsonFile.js:
```diff
@@ -0,0 +1,82 @@
+
+function filterProperties(listProps, object) {
+  const filtered = {}
+  listProps.map(property => {
```
Using `map` here doesn't make sense if you're not doing something with what's returned. Either use `forEach` or `reduce`

MacKentoch replied:
Done

kentcdodds commented on generate/writeToJsonFile.js:
```diff
@@ -0,0 +1,82 @@
+    const allFileContent = parseReadData(data)
+
+    // UPDATE MATCHING ARRAY ITEMS (ONLY) TO CONTENT
+    arrayNewItems.map(item => {
```
Same here as below:

> Using `map` here doesn't make sense if you're not doing something with what's returned. Either use `forEach`

MacKentoch replied:
Yes you are right. Sorry if you had to spend time on that PR it was not the objective.

kentcdodds commented:
Overall, I think it's pretty good. I think that it's a little overcomplicated by trying to only update the json files rather than just entirely regenerating them every time. Particularly when this is run on Travis, those files wont ever exist.

MacKentoch commented:
It should be better now since really simplified `generate/writeToJsonFile.js` (_more readable_).

Sorry again if you lost time on that PR.

kentcdodds commented on generate/jsonFiles.js:
```diff
@@ -0,0 +1,29 @@
+import path from 'path'
+import {writeJsonArray, writeJsonObject} from './writeToJsonFile'
```
Oh, sorry, I totally forgot to mention, could you name all files with `kebab-case` rather than `camelCase` to avoid issues with `git` file renames and other cross-platform problems?

kentcdodds replied:
That's files and directories should all be kebab case

MacKentoch replied:
Sure no problem.

kentcdodds commented on generate/writeToJsonFile.js:
```diff
@@ -0,0 +1,45 @@
+
+const JsonStringifyParameters = {
+  replacer: null,
+  space: 4,
```
Could we make this `2` please?

MacKentoch replied:
Sorry I did not understand what were these 2. 
Feel free to tell me after latest PR update.

kentcdodds commented on generate/json-files-configs/episodes.js:
```diff
@@ -0,0 +1,15 @@
+export const episodesToJson = {
+  destFileName: 'data',
+  destDirectoryFromHere: '../episodes/',
+  testEqualityBy: 'date',
```
Looks like `testEqualityBy` is no longer necessary.

MacKentoch replied:
Yes you are right. 
Done.

kentcdodds commented on generate/write-to-json-file.js:
```diff
@@ -0,0 +1,45 @@
+}
+
+export function writeJsonArray(arrayNewItems, config, applyFilter = false) {
+  const DEST_JSON_FILE = path.join(__dirname, config.destDirectoryFromHere, `${config.destFileName}.json`)
```
Looks like all `destFileName`s are `data`. Could we just remove that property from the config and hard code it as `data.json` here?

MacKentoch replied:
It is your repo (your baby) so I just follow what tell me without hesitation.
`data.json` is now hard coded and removed from configs.

kentcdodds commented on generate/write-to-json-file.js:
```diff
@@ -0,0 +1,45 @@
+    replacer,
+    space,
+  } = JsonStringifyParameters
+  return JSON.stringify(thatObject, replacer, space)
```
Could we just hard code things here? I don't see the reason to have a `JsonStringifyParameters` object. In fact, I don't really see any value in having this function.

MacKentoch replied:
Removed.

kentcdodds commented on generate/write-to-json-file.js:
```diff
@@ -0,0 +1,45 @@
+export function writeJsonObject(newItem, config, applyFilter = false) {
+  const DEST_JSON_FILE = path.join(__dirname, config.destDirectoryFromHere, `${config.destFileName}.json`)
+  const itemToCopy = applyFilter ? filterProperties(config.propertiesToFilter, newItem) : newItem
+  writeToFile(DEST_JSON_FILE, stringify(itemToCopy))
```
Could this just be:

``` javascript
writeToFile(DEST_JSON_FILE, JSON.stringify(itemToCopy))
```

I don't think we need to do any spacing of any kind here...

MacKentoch replied:
By default, JSON is written with no space and on a single line. 
The reason behind : in fact it is just because it breaks my Atom when I read an episode json file (_single episode json: there is no parameter filter, so, with transcripts it can fast become a super huge line_).

**Updated to simple:** 

``` javascript
writeToFile(DEST_JSON_FILE, JSON.stringify(itemToCopy))
```

kentcdodds commented on generate/write-to-json-file.js:
```diff
@@ -0,0 +1,45 @@
+  const allFileContent = arrayNewItems.map(item => {
+    return applyFilter ? filterProperties(config.propertiesToFilter, item) : item
+  })
+  writeToFile(DEST_JSON_FILE, stringify(allFileContent))
```
Same thing here. Just use `JSON.stringify` directly.

MacKentoch replied:
Related to previous and done that way now.

kentcdodds commented on generate/write-to-json-file.js:
```diff
@@ -0,0 +1,45 @@
+function writeToFile(file, content) {
+  fs.writeFile(file, content, err => {
+    if (err) {
+      process.exit(1)
```
If we must `exit` here, could we at least log the error?

MacKentoch replied:
Sure, I log now through `process.stdout` : 

``` javascript
process.stdout.write(`😱ERROR: could not write json file: ${file}\n->Error details: ${err}\n`)
```

kentcdodds commented:
Hey, thanks for working with me on this PR :-) I'm excited to have this!

MacKentoch commented:
Nice working with you on this PR.

If you knew what is coming next you would be even more excited (_like I am: I feel impatient to tell you more about but time has not come right now_).

kentcdodds commented on generate/write-to-json-file.js:
```diff
@@ -0,0 +1,33 @@
+function writeToFile(file, content) {
+  fs.writeFile(file, content, err => {
+    if (err) {
+      process.stdout.write(`😱ERROR: could not write json file: ${file}\n->Error details: ${err}\n`)
```
Thanks for the emoji :-)

kentcdodds commented:
I'm super excited to know what you're working on! Do you think you could do one other thing? Add `createJson` to the `default` `build` script? That way it's built automatically with the rest of everything when the site is built and deployed. That'll be it! Thanks!

MacKentoch commented:
Sorry I forgot to add it in build script.
Now it is updated (by the way thanks for `p-s` it is pretty handy).

kentcdodds commented on package-scripts.js:
```diff
@@ -7,7 +7,7 @@ module.exports = {
       default: {
         description: 'Runs all the build scripts in parallel',
-        script: 'p-s -p eslint,build.home,build.episodes,build.guests,build.deals,build.css,build.contributors',
+        script: 'p-s -p eslint,build.home,build.episodes,build.guests,build.deals,build.css,build.contributors,createJson.script',
```
Should just be `createJson` please leave of the `.script`

kentcdodds commented:
Oh! And don't forget to add your profile photo! There's a script for that.

MacKentoch commented:
I'd like to but I think I did not understand the script: 

``` bash
npm start addContributor
```

returns

``` bash
p-s executing: babel-node ./other/add-contributor
/Users/kentor/GitProjects/jsairWebsite/site/other/utils/twitter.js:48
    throw new Error('you must provide Twitter API info with a twitter.api.ignored.json');
    ^

Error: you must provide Twitter API info with a twitter.api.ignored.json
```

kentcdodds commented:
Oh yeah! Totally forgot about that... Hmmm... OK, I'll take care of it. Thanks!

MacKentoch commented:
Yes🍸!

Thank you @kentcdodds !

Let's finish my little stuff.
